### ISSUE TITLE
Templating importent classes

### ISSUE THREAD

**Issue by benjaminhuth** [lwtnn/lwtnn#102]
Hello,

It would be very useful if one could choose the underlying data-type which lwtnn uses. E.g., one could use single precision numbers, or even differentiable types. I tried this in [a fork of lwtnn](https://github.com/benjaminhuth/lwtnn).

As a starting point for the discussion, the following snippet shows how I approached this: 

- LWTNN right now:
   ```c++
    class Stack
    {
        // something with double...
    };
   ```
   
- Proposal:   
   ```c++
    template<typename T>
    class StackT
    {
        // something generic
    };
    
    using Stack = StackT<double>;
   ```
The typedefs are introduced to maintain compatibility with the rest of the library. 

I also think, the interface-classes can stay unchanged, we could ensure with a ``static_assert(...)`` that the type ``T`` can be assigned with double/converted to double. 

I would propose this as a change, but maybe there are also other approaches out there which we can disucss.

Best regards,
Benjamin

**Comment by dguest** [comment]
Hi,

Thanks for making an issue, and sorry for the slow response. I guess the only thing I'm unsure of is the naming convention, but I don't know of any more standard ways to deal with this.

**Comment by lukasheinrich** [comment]
@dguest you mean `StackT` and `Stack`?

**Comment by dguest** [comment]
@lukasheinrich yeah. I guess most people would be doing this the other way around: the template would have the shorter name and the `Stack` would be `StackD` or something. But obviously we can't do that without breaking the existing code. So if no one has a better idea I'm fine with this.

**Comment by dguest** [comment]
I'm also wondering if there is any reasonable way to hide the 

```C++
#include <Eigen/Dense>
```

From e.g. the `LightweightNeuralNetwork.hh` header with some kind of pImpl tricks. In the current code most of the ATLAS facing code doesn't include all the headers.

I can think of some rather convoluted solutions like having the template classes inherit from a base class, not sure if anyone has a better idea.

**Comment by benjaminhuth** [comment]
Yeah the problem with the templated classes is that you always have to include the whole implementation. Maybe something like the following? (I think this is in fact pImpl).

```c++
// MyClassT.hpp
#include <Eigen/Dense>

template<class T>
class MyClassT
{
    public: 
        void do_something(T t) { /* use Eigen ... */ }
};
```
```c++
// MyClass.hpp

template<class T> class MyClassT;

class MyClass
{
    public: 
        void do_something(double t);
    private: 
        std::unique_ptr<MyClassT<double>> generic_class;
};
```
```c++
// MyClass.cpp
#include "MyClass.hpp"
#include "MyClassT.hpp"

void MyClass::do_something(double t)
{
    generic_class->do_something(t);
}
```
That would do the trick for double I think.

Of course, if you want do use the generic version, you still have the dependency to Eigen... but I think there is no way to get around this.

What I'm not sure about is the runtime overhead of this approach.

**Comment by dguest** [comment]
I didn't realize this kind if forward template declaration worked, if that works I'm all for it (and we'd only need it for double, this is just for backward compatibility)

**Comment by benjaminhuth** [comment]
I was able to compile the above example, so I think it works. I will try to do this in my LWTNN fork, then we can have a look at it.

**Comment by lukasheinrich** [comment]
yeah that's pretty pimpl-y

**Comment by benjaminhuth** [comment]
I tried the pImpl thing for the NeuralNetwork classes in my fork, you can have a look at it. It's in the branch "feature/templated".
It compiles, and as far I see, it works like expected. But there are some new compiler warnings which I've not investigated yet...

**Comment by dguest** [comment]
Good that it complies! If you're pretty close to something we can merge feel free to make a WIP PR.

On that subject I guess there are a few things to discuss: 
 - The templating is fine and doesn't break any backward compatibility
 - The training part should be considered carefully (we don't have a lot of concrete use cases yet and it adds some complexity)
 - I just remembered that C++17 isn't supported in the branch of ATLAS code where most developments are going (the 21.2 branch still uses gcc 6.2 in some builds).

I guess these aren't on topic for this issue but something to be aware of.

**Comment by benjaminhuth** [comment]
Hi all,

Sorry that it took so long to implement these things. But now I think most part of the work is done. During the process I thought again about the naming conventions, and now I'm not sure anymore if its the perfect choice. So what I did now is something like that:

```c++
namespace lwt {
namespace generic {
  
  template<typename T>
  class Stack
  {
    // ...
  };

} // namespace generic

using Stack = generic::Stack<double>;

} // namespace lwt
```

Besides that, I separated the files in a sub-folder 'generic' in the include directory. At least to me this is a bit more expressive... One could rename 'generic' also to 'detail' or something like that... But if you think that the style we discussed before is better suited it's not to hard to revert this back...

The code is at my fork, in the branch "feature/templated". In this branch are just changes for template-change, nothing related to training.

**Comment by dguest** [comment]
I like the namespace idea. Feel free to make a WIP merge request: it sounds like everything you've done will add a lot of flexibility without breaking the interface or exposing any additional code to the current headers, so unless you have more questions we could probably start the final review.

**Comment by dguest** [comment]
Ok, I just merged #103. Thanks for all the work! I had a few followup points:
 - There was a leftover `friend class TrainableLWTNN` declaration in a few places. I'm not really opposed to keeping it there if you are using it for something, but if it's not needed I'd prefer to delete it.
 - @benjaminhuth, if you want to add yourself to a contributor please add your name, ORCID, affiliation etc to the `.zenodo.json` file and make another merge request. You've certainly done enough work at this point but it's up to you.

**Comment by benjaminhuth** [comment]
Cool, I'm happy its done! I also created a PR because of the contribution list. These `friend` declarations can be deleted, I should have checked myself for such leftovers...
